server/EXCEPTIONS.md

```diff
@@ -0,0 +1,168 @@
+# Exception handling in our backend
+
+## Quickstart
+
+To generate errors in the backend that will be passed to the
+frontend, throw a [Nest-defined HttpException](https://docs.nestjs.com/exception-filters#throwing-standard-exceptions), <u>_generally within a service_</u>:
+
+```js
+  throw new HttpException(responseBody, status)
+```
+
+Where...
+  - `status` is one of `HttpStatus` enum codes. E.g. `HttpStatus.UNAUTHORIZED`
+  - `responseBody` argument is an object with the following schema.
+    ```js
+       const responseBody = {
+        code: 'APPLICANT_PORTAL_ERROR_CODE',
+        message: 'Applicant Portal could not find coffee. CRM: CRM ran out of tea.',
+      }
+    ```
+    - The `code` property is developer-defined, and should be a unique, concise,
+    descriptive identifier for the type of error. Users can provide this code to
+    help developers locate the code that yielded the error.
+    - The `message` property is a short, human-readable sentence firstly
+    describing the error for the benefit of end users. Secondly, it should contain
+    errors relayed from CRM, if available.
+
+To scope the error to a specific part of the code, use try-catch blocks. For example,
+the following try-catch scopes the thrown error to an attempt to find a project.
+
+```js
+// services/my-project.service.ts
+getProject(){
+  try {
+    return myProject = this.get('dcp_project', ...);
+  } catch (e) {
+    throw new HttpException({
+      code: 'GET_PROJECTS',
+      message: `Failed to find a project for given id. CRM: ${e}`,
+    });
+  }
+}
+```
+
+Generally, when generating errors as described above, <u>*do so within Nest services*</u>.
+
+## Relaying Errors
+
+In the controller that consumes a service, _relay_ the error from the service by simply throwing the error as it bubbles up:
+
+```js
+// controllers/my-project.controller.ts
+@Get('/')
+  async getProject(@Session() session) {
+    try {
+      return this.myProjectService.getProject();
+    } catch (e) {
+      throw e; // simply relay the bubbling error
+    }
+  }
+}
+```
+
+Note how it is not necessary to construct a new HttpException in the controller to relay the error from the service. Instead, we simply use `throw` to pass along the error `e` from the service.
+
+Errors will *bubble* through the function call stack. The `catch` block in a try-catch will also catch errors generated from *within* functions that are called in the `try` block.
+
+Only the HttpException caught and thrown closest to or on the controller
+endpoint will be returned for a given request. To prevent confusion over which
+backend layer should provide the most detailed error handling, this document prescribes generating new HTTPExceptions in services, and simply relaying them in controllers.
+
+## General Errors / Preventing Gaps
+Since we are trying to `throw new HttpException` at the service level, we may
+want to write a generic, controller-level catch-all for errors. This will help catch
```

Comment:
okay like an even bigger safety net

Reply:
> I guess if we want to know more specifics we have to be reallyyy diligent about wrapping most stuff in the services in try-catch blocks
Yes, I think this is the case. But curious about thoughts from @allthesignals 
>  is there another strategy for making sure errors don't seep through services uncaught?
Yep, I think this is an open question!